## Segmenting and Clustering Neighborhoods in Toronto (Part 3)

This notebook consists of answer for Part 1, 2 and 3 of the Assignment.

## Part 1

In [1]:
!pip install bs4
print('\nBeautifulSoup package installation success.')


BeautifulSoup package installation success.


Import libraries

In [2]:
from bs4 import BeautifulSoup 
import requests
import pandas as pd

The following url contains html tables with list of postal codes in Canada where the first letter is M. Postal codes beginning with M are located within the city of Toronto.

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

Read the HTML tables inside the Wikipedia page.

In [4]:
table_data = pd.read_html(url, flavor='bs4')
print(f'Total tables: {len(table_data)}')

Total tables: 3


In [5]:
table_data[0].head()

0                1  \
0                                    M1ANot assigned  M2ANot assigned   
1                    M1BScarborough(Malvern / Rouge)  M2BNot assigned   
2  M1CScarborough(Rouge Hill / Port Union / Highl...  M2CNot assigned   
3  M1EScarborough(Guildwood / Morningside / West ...  M2ENot assigned   
4                             M1GScarborough(Woburn)  M2GNot assigned   

                                                2  \
0                        M3ANorth York(Parkwoods)   
1                   M3BNorth York(Don Mills)North   
2  M3CNorth York(Don Mills)South(Flemingdon Park)   
3                                 M3ENot assigned   
4                                 M3GNot assigned   

                                                3  \
0                 M4ANorth York(Victoria Village)   
1  M4BEast York(Parkview Hill / Woodbine Gardens)   
2                  M4CEast York(Woodbine Heights)   
3                    M4EEast Toronto(The Beaches)   
4                           M4GEast York(Leaside)   

                                                 4  \
0  M5ADowntown Toronto(Regent Park / Harbourfront)   
1    M5BDowntown Toronto(Garden District, Ryerson)   
2              M5CDowntown Toronto(St. James Town)   
3                 M5EDowntown Toronto(Berczy Park)   
4          M5GDowntown Toronto(Central Bay Street)   

                                                  5  \
0  M6ANorth York(Lawrence Manor / Lawrence Heights)   
1                          M6BNorth York(Glencairn)   
2                       M6CYork(Humewood-Cedarvale)   
3                      M6EYork(Caledonia-Fairbanks)   
4                     M6GDowntown Toronto(Christie)   

                                                6                7  \
0  M7AQueen's Park(Ontario Provincial Government)  M8ANot assigned   
1                                 M7BNot assigned  M8BNot assigned   
2                                 M7CNot assigned  M8CNot assigned   
3                                 M7ENot assigned  M8ENot assigned   
4                                 M7GNot assigned  M8GNot assigned   

                                                   8  
0                     M9AEtobicoke(Islington Avenue)  
1  M9BEtobicoke(West Deane Park / Princess Garden...  
2  M9CEtobicoke(Eringate / Bloordale Gardens / Ol...  
3                                    M9ENot assigned  
4                                    M9GNot assigned

We want to capture Postal codes, Borough and Neighborhood data from the first table.

Use BeautifulSoup package to transform the Wikipedia data into dataframe.

In [6]:
url_data  = requests.get(url).text
soup = BeautifulSoup(url_data,"html5lib")
tables = soup.find_all('table')
table_data = pd.DataFrame(columns=['PostalCode', 'Borough', 'Neighborhood'])

for row in tables[0].tbody.find_all('td'):
    if row.span.text == 'Not assigned':
        pass
    else:
        postalcode = row.p.text[:3] 
        borough = (row.span.text).split('(')[0]
        neighborhood = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_data = table_data.append({"PostalCode" : postalcode, "Borough" : borough, "Neighborhood" : neighborhood}, ignore_index=True)

table_data

PostalCode                                            Borough  \
0          M3A                                         North York   
1          M4A                                         North York   
2          M5A                                   Downtown Toronto   
3          M6A                                         North York   
4          M7A                                       Queen's Park   
..         ...                                                ...   
98         M8X                                          Etobicoke   
99         M4Y                                   Downtown Toronto   
100        M7Y  East TorontoBusiness reply mail Processing Cen...   
101        M8Y                                          Etobicoke   
102        M8Z                                          Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [7]:
table_df = pd.DataFrame(table_data)
table_df['Borough'] = table_df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
table_df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Make sure "Not assigned" neighborhood is not inside dataframe.

In [8]:
table_df.Neighborhood.str.count('Not assigned').sum()

0

In [9]:
len(table_df['PostalCode'].unique())

103

There are 103 unique postal code inside the dataframe.

In [10]:
table_df.shape

(103, 3)

### Answer: There are 103 rows inside the dataframe.

---

## Part 2

In [11]:
!pip install geocoder
print('\nGeocoder package installation success.')


Geocoder package installation success.


Get the latitude and longitude coordinates of each neighborhood in Toronto.

In [12]:
import geocoder

geo_coords = None
postal_code = 'M5G'
count = 0

while (geo_coords is None and count < 2500):
    geo = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
    count = count + 1
    geo_coords = geo.latlng
    
if (geo_coords is not None):
    latitude = geo_coords[0]
    longitude = geo_coords[1]
    print(latitude, longitude)
else:
    print('Try ' + str(count) + ' times to get the geographical coordinates through geocoder but not success.')

Try 2500 times to get the geographical coordinates through geocoder but not success.


Not able to get the geographical coordinates of the neighborhoods using the Geocoder package. Get the geographical coordinates for each postal code from csv file.

In [13]:
coordinate_data = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv')
coordinate_data.columns = ['PostalCode', 'Latitude', 'Longitude']
coordinate_data.sort_values(by=['PostalCode'], ascending=True, inplace=True)
coordinate_data

PostalCode   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
..         ...        ...        ...
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

[103 rows x 3 columns]

Sorting ascending on table_contents data.

In [14]:
table_df.sort_values(by=['PostalCode'], ascending=True, inplace=True, ignore_index=True)
table_df

PostalCode              Borough  \
0          M1B          Scarborough   
1          M1C          Scarborough   
2          M1E          Scarborough   
3          M1G          Scarborough   
4          M1H          Scarborough   
..         ...                  ...   
98         M9N                 York   
99         M9P            Etobicoke   
100        M9R            Etobicoke   
101        M9V            Etobicoke   
102        M9W  Etobicoke Northwest   

                                          Neighborhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Port Union, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
..                                                 ...  
98                                              Weston  
99                                           Westmount  
100  Kingsview Village, St. Phillips, Martin Grove ...  
101  South Steeles, Silverstone, Humbergate, Jamest...  
102  Clairville, Humberwood, Woodbine Downs, West H...  

[103 rows x 3 columns]

Merge the table_df and coordinate_data dataframe together.

In [15]:
toronto_df = pd.merge(table_df, coordinate_data, how="left", on=["PostalCode"])
toronto_df

PostalCode              Borough  \
0          M1B          Scarborough   
1          M1C          Scarborough   
2          M1E          Scarborough   
3          M1G          Scarborough   
4          M1H          Scarborough   
..         ...                  ...   
98         M9N                 York   
99         M9P            Etobicoke   
100        M9R            Etobicoke   
101        M9V            Etobicoke   
102        M9W  Etobicoke Northwest   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724  
101  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
102  Clairville, Humberwood, Woodbine Downs, West H...  43.706748 -79.594054  

[103 rows x 5 columns]

In [16]:
toronto_df.shape

(103, 5)

We get 103 records as we execute the left merge method on 'PosalCode' key.

---

### Part 3

Cluster the neighborhoods of Toronto based on similarity of the venues categories by using k-means clustering and foursquare API.

In [17]:
!conda install -c conda-forge geopy --yes 
!conda install -c conda-forge folium=0.5.0 --yes 

import folium                              # map rendering library
import json                                # library to handle JSON files
from geopy.geocoders import Nominatim      # convert an address into latitude and longitude values
from pandas.io.json import json_normalize  # tranform JSON file into a pandas dataframe

pd.set_option('display.max_columns', None) 
pd.set_option('display.max_rows', None)

Solving environment: ...working... done

# All requested packages already installed.

Solving environment: ...working... done

# All requested packages already installed.



In [18]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent='toronto_explorer')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The goegrapical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The goegrapical coordinate of Toronto City are 43.6534817, -79.3839347.


Create a map of Toronto with borough and neighborhood using latitude and longitude values.

In [19]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, lng, borough, neighborhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat, lng],
    radius=5,
    popup=label,
    color='green',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(map_toronto)
    
map_toronto

Utilize Foursquare API to explore the neighborhoods and segment them.

In [20]:
CLIENT_ID = 'AR1G3JHCYDSTEUQRVGV212ZRWWUSASQYJVFEXD5DSHQ5LKB3' # Foursquare ID
CLIENT_SECRET = 'E3OULZT01GLEFCGRWRC4NJ43UQ1RSIZTM5DG2W5C255RLFWB' # Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

### Explore Neighborhoods in Toronto

Generalize to get venues for each neighborhood in Toronto

In [21]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [22]:
toronto_venues = getNearbyVenues(names=toronto_df['Neighborhood'], latitudes=toronto_df['Latitude'], longitudes=toronto_df['Longitude'])

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Don Mills South
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
The Danforth  East
The Danforth West, Riverdale


In [23]:
print(toronto_venues.shape)
toronto_venues.head()

(2141, 7)


Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                      Venue  Venue Latitude  \
0              -79.194353                    Wendy’s       43.807448   
1              -79.160497      Royal Canadian Legion       43.782533   
2              -79.160497  Affordable Toronto Movers       43.787919   
3              -79.188711             RBC Royal Bank       43.766790   
4              -79.188711          G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.162977         Moving Target  
3       -79.191151                  Bank  
4       -79.191537     Electronics Store

There are 2141 of venues at Toronto neighborhoods.

In [24]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                  8   
Bathurst Manor, Wilson Heights, Downsview North                        23   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      24   
Berczy Park                                                            56   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks                                                     4   
Cedarbrae                                                               9   
Central Bay Street                                                     66   
Christie                                                               16   
Church and Wellesley                                                   80   
Clairville, Humberwood, Woodbine Downs, West Hu...                      4   
Clarks Corners, Tam O'Shanter, Sullivan                                13   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        5   
Don Mills North                                                         4   
Don Mills South                                                        20   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Downsview Central                                                       3   
Downsview East                                                          2   
Downsview Northwest                                                     4   
Downsview West                                                          6   
Dufferin, Dovercourt Village                                           16   
Enclave of L4W                                                         13   
Enclave of M4L                                                         16   
Enclave of M5E                                                         97   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           65   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Glencairn                                                               5   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       9   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
Hillcrest Village                                                       6   
Humber Summit                                                           1   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      3   
India Bazaar, The Beaches West                                         21   
Kennedy Park, Ionview, East Birchmount Park                             6   
Kensington Market, Chinatown, Grange Park                              66   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                

### Analyze Each Neighborhood

One hot encoding the venue category.

In [25]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

print(toronto_onehot.shape)
toronto_onehot.head()

(2141, 271)


Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                  0               0        0                   0   
1                  0               0        0                   0   
2                  0               0        0                   0   
3                  0               0        0                   0   
4                  0               0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    1   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     1    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0               0                 0                   0      0   
1               0                 0                   0      0   
2               0                 0                   0      0   
3               0                 0                   0      0   
4               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Business Service  \
0             0              0       

There are 271 different types of venue categories.

Add neighborhood column to dataframe.

In [26]:
toronto_onehot['Neighborhoods'] = toronto_venues['Neighborhood'] 
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Neighborhoods  Accessories Store  Adult Boutique  \
0                          Malvern, Rouge                  0               0   
1  Rouge Hill, Port Union, Highland Creek                  0               0   
2  Rouge Hill, Port Union, Highland Creek                  0               0   
3       Guildwood, Morningside, West Hill                  0               0   
4       Guildwood, Morningside, West Hill                  0               0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0     0    0               0                 0                   0      0   
1     0    1               0                 0                   0      0   
2     0    0               0                 0                   0      0   
3     1    0               0                 0                   0      0   
4     0    0               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            

In [27]:
toronto_grouped = toronto_onehot.groupby('Neighborhoods').mean().reset_index()
toronto_grouped.head()

Neighborhoods  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0              0.0               0.0             0.000000           0.0   
1              0.0               0.0             0.000000           0.0   
2              0.0               0.0             0.000000           0.0   
3              0.0               0.0             0.000000           0.0   
4              0.0               0.0             0.041667           0.0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0       0.0          0.0         0.0                  0.0               0.0   
1       0.0          0.0         0.0                  0.0               0.0   
2       0.0          0.0         0.0                  0.0               0.0   
3       0.0          0.0         0.0                  0.0               0.0   
4       0.0          0.0         0.0                  0.0               0.0   

   Athletics & Sports  Auto Garage  Auto Workshop  BBQ Joint  Baby Store  \
0                 0.0          0.0            0.0        0.0         0.0   
1                 0.0          0.0            0.0        0.0         0.0   
2                 0.0          0.0            0.0        0.0         0.0   
3                 0.0          0.0            0.0        0.0         0.0   
4                 0.0          0.0            0.0        0.0         0.0   

   Bagel Shop  Bakery      Bank  Bar  Baseball Field  Baseball Stadium  \
0         0.0     0.0  0.000000  0.0             0.0               0.0   
1         0.0     0.0  0.000000  0.0             0.0               0.0   
2         0.0     0.0  0.086957  0.0             0.0               0.0   
3         0.0     0.0  0.250000  0.0             0.0               0.0   
4         0.0     0.0  0.000000  0.0             0.0               0.0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                 0.0    0.0              0.0       0.0         0.0   
1                 0.0    0.0              0.0       0.0         0.0   
2                 0.0    0.0              0.0       0.0         0.0   
3                 0.0    0.0              0.0       0.0         0.0   
4                 0.0    0.0              0.0       0.0         0.0   

   Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                 0.0        0.0     0.0            0.0        0.0       0.0   
1                 0.0        0.0     0.0            0.0        0.0       0.0   
2                 0.0        0.0     0.0            0.0        0.0       0.0   
3                 0.0        0.0     0.0            0.0        0.0       0.0   
4                 0.0        0.0     0.0            0.0        0.0       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                   0.0            0.25      0.0     0.000000   
1                   0.0            0.00      0.0     0.000000   
2                   0.0            0.00      0.0     0.043478   
3                   0.0            0.00      0.0     0.000000   
4                   0.0            0.00      0.0     0.0

Function to get the most common venues.

In [28]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create new dataframe and display top 10 venues for each neighborhood in Toronto.

In [29]:
# import numpy to handle data in a vectorized manner
import numpy as np

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)

neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhoods']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village        Japanese Restaurant   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Clothing Store        Breakfast Spot                Lounge   
1              Pharmacy                  Pool           Coffee Shop   
2           Coffee Shop           Gas Station      Sushi Restaurant   
3    Chinese Restaurant                  Bank                  Café   
4           Coffee Shop    Italian Restaurant          Cupcake Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Motel   Moroccan Restaurant   Monument / Landmark   
1        Sandwich Place                   Pub                   Gym   
2           Supermarket    Frozen Yogurt Shop   Fried Chicken Joint   
3     Accessories Store                 Motel   Moroccan Restaurant   
4      Toy / Game Store               Butcher                  Café   

             8th Most Common Venue            9th Most Common Venue  \
0  Molecular Gastronomy Restaurant       Modern European Restaurant   
1                    Metro Station               Mexican Restaurant   
2                   Sandwich Place                       Restaurant   
3              Monument / Landmark  Molecular Gastronomy Restaurant   
4                     Liquor Store                 Sushi Restaurant   

       10th Most Common Venue  
0           Mobile Phone Shop  
1   Middle Eastern Restaurant  
2                       Diner  
3  Modern European Restaurant  
4                   Juice Bar

### Cluster Toronto Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [30]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhoods', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0])

Create new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [31]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

# merge neighborhoods_venues_sorted with toronto_df to add Postal Code, Borough, Latitude and Longitude for each neighborhood
toronto_merged = pd.merge(toronto_df, neighborhoods_venues_sorted, how="inner", on=["Neighborhood"])

toronto_merged.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0 -79.194353               4  Fast Food Restaurant      Accessories Store   
1 -79.160497               1         Moving Target                    Bar   
2 -79.188711               1   Rental Car Location         Breakfast Spot   
3 -79.216917               1           Coffee Shop  Korean BBQ Restaurant   
4 -79.239476               1                Bakery        Thai Restaurant   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0  Middle Eastern Restaurant   Moroccan Restaurant   Monument / Landmark   
1         Miscellaneous Shop   Moroccan Restaurant   Monument / Landmark   
2             Medical Center                  Bank    Mexican Restaurant   
3          Accessories Store    Miscellaneous Shop   Moroccan Restaurant   
4                Gas Station                Lounge      Hakka Restaurant   

             6th Most Common Venue            7th Most Common Venue  \
0  Molecular Gastronomy Restaurant       Modern European Restaurant   
1  Molecular Gastronomy Restaurant       Modern European Restaurant   
2                       Donut Shop                     Intersection   
3              Monument / Landmark  Molecular Gastronomy Restaurant   
4             Caribbean Restaurant               Athletics & Sports   

        8th Most Common Venue 9th Most Common Venue  \
0           Mobile Phone Shop    Miscellaneous Shop   
1           Mobile Phone Shop     Accessories Store   
2           Electronics Store            Restaurant   
3  Modern European Restaurant     Mobile Phone Shop   
4         Fried Chicken Joint                  Bank   

            10th Most Common Venue  
0               Mexican Restaurant  
1                            Motel  
2       Modern European Restaurant  
3        Middle Eastern Restaurant  
4  Molecular Gastronomy Restaurant

Plot the clusters on the map.

In [32]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm 
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters of Toronto

Cluster 1

In [33]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19        North York               0                  Park   
21        North York               0                  Park   
28        North York               0               Airport   
48  Downtown Toronto               0                  Park   
72              York               0                  Park   
95              York               0                  Park   

   2nd Most Common Venue      3rd Most Common Venue  \
19     Accessories Store  Middle Eastern Restaurant   
21     Convenience Store          Accessories Store   
28                  Park          Accessories Store   
48            Playground                      Trail   
72         Women's Store                       Pool   
95     Accessories Store  Middle Eastern Restaurant   

        4th Most Common Venue      5th Most Common Venue  \
19        Moroccan Restaurant        Monument / Landmark   
21  Middle Eastern Restaurant        Moroccan Restaurant   
28  Middle Eastern Restaurant        Moroccan Restaurant   
48         Miscellaneous Shop        Moroccan Restaurant   
72          Accessories Store  Middle Eastern Restaurant   
95        Moroccan Restaurant        Monument / Landmark   

              6th Most Common Venue            7th Most Common Venue  \
19  Molecular Gastronomy Restaurant       Modern European Restaurant   
21              Monument / Landmark  Molecular Gastronomy Restaurant   
28              Monument / Landmark  Molecular Gastronomy Restaurant   
48              Monument / Landmark  Molecular Gastronomy Restaurant   
72              Monument / Landmark  Molecular Gastronomy Restaurant   
95  Molecular Gastronomy Restaurant       Modern European Restaurant   

         8th Most Common Venue 9th Most Common Venue  \
19           Mobile Phone Shop    Miscellaneous Shop   
21  Modern European Restaurant     Mobile Phone Shop   
28  Modern European Restaurant     Mobile Phone Shop   
48  Modern European Restaurant     Mobile Phone Shop   
72  Modern European Restaurant     Mobile Phone Shop   
95           Mobile Phone Shop    Miscellaneous Shop   

       10th Most Common Venue  
19         Mexican Restaurant  
21         Miscellaneous Shop  
28         Miscellaneous Shop  
48  Middle Eastern Restaurant  
72         Miscellaneous Shop  
95         Mexican Restaurant

Cluster 2

In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5,toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1              Scarborough               1              Moving Target   
2              Scarborough               1        Rental Car Location   
3              Scarborough               1                Coffee Shop   
4              Scarborough               1                     Bakery   
5              Scarborough               1                 Playground   
6              Scarborough               1             Discount Store   
8              Scarborough               1        American Restaurant   
9              Scarborough               1            College Stadium   
10             Scarborough               1          Indian Restaurant   
12             Scarborough               1  Latin American Restaurant   
13             Scarborough               1                Pizza Place   
15             Scarborough               1                Coffee Shop   
16              North York               1       Fast Food Restaurant   
17              North York               1             Clothing Store   
18              North York               1        Japanese Restaurant   
20              North York               1           Ramen Restaurant   
22              North York               1                Pizza Place   
24              North York               1       Caribbean Restaurant   
25              North York               1                        Gym   
26              North York               1                       Bank   
27              North York               1     Furniture / Home Store   
30              North York               1             Baseball Field   
31              North York               1               Liquor Store   
32              North York               1                Pizza Place   
33               East York               1                Pizza Place   
35            East Toronto               1          Health Food Store   
36               East York               1        Sporting Goods Shop   
37               East York               1          Indian Restaurant   
39            East Toronto               1           Greek Restaurant   
40            East Toronto               1             Sandwich Place   
41            East Toronto               1                Coffee Shop   
44         Central Toronto               1                Coffee Shop   
45         Central Toronto               1               Dessert Shop   
47         Central Toronto               1                Coffee Shop   
49        Downtown Toronto               1                       Café   
50        Downtown Toronto               1                Coffee Shop   
51        Downtown Toronto               1                Coffee Shop   
52        Downtown Toronto               1                Coffee Shop   
53        Downtown Toronto               1                Coffee Shop   
54        Downtown Toronto               1                     Bakery   
55        Downtown Toronto               1                Coffee Shop   
56        Downtown Toronto               1                Coffee Shop   
57        Downtown Toronto               1                Coffee Shop   
58        Downtown Toronto               1                Coffee Shop   
59        Downtown Toronto               1                Coffee Shop   
60              North York               1             Sandwich Place   
63         Central Toronto               1             Sandwich Place   
64        Downtown Toronto               1                       Café   
65        Downtown Toronto               1                        Bar   
66        Downtown Toronto               1            Airport Service   
67  Downtown Toronto Stn A               1                Coffee Shop   
68        Downtown Toronto               1                Coffee Shop   
69              North York               1             Clothing Store   
71                    York               1                      Field   
73  

Cluster 3

In [35]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
46  Central Toronto               2                  Park   
93       North York               2                   Gym   

   2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
46                   Gym                      Trail     Accessories Store   
93     Accessories Store  Middle Eastern Restaurant   Moroccan Restaurant   

        5th Most Common Venue            6th Most Common Venue  \
46  Middle Eastern Restaurant              Monument / Landmark   
93        Monument / Landmark  Molecular Gastronomy Restaurant   

              7th Most Common Venue       8th Most Common Venue  \
46  Molecular Gastronomy Restaurant  Modern European Restaurant   
93       Modern European Restaurant           Mobile Phone Shop   

   9th Most Common Venue 10th Most Common Venue  
46     Mobile Phone Shop     Miscellaneous Shop  
93    Miscellaneous Shop     Mexican Restaurant

Cluster 4

In [36]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
7              Scarborough               3                   Bus Line   
11             Scarborough               3                     Bakery   
14             Scarborough               3                 Playground   
23              North York               3       Fast Food Restaurant   
29              North York               3              Grocery Store   
34               East York               3               Skating Rink   
38  East York/East Toronto               3               Intersection   
42         Central Toronto               3                       Park   
43         Central Toronto               3             Breakfast Spot   
61         Central Toronto               3       Fast Food Restaurant   
62         Central Toronto               3           Sushi Restaurant   
70              North York               3         Italian Restaurant   
77              North York               3             Massage Studio   
88               Etobicoke               3                       Park   
91               Etobicoke               3  Middle Eastern Restaurant   
97               Etobicoke               3             Sandwich Place   

         2nd Most Common Venue  3rd Most Common Venue  \
7                       Bakery            Bus Station   
11                 Auto Garage  Vietnamese Restaurant   
14                        Park           Intersection   
23                        Park      Food & Drink Shop   
29                        Bank                   Park   
34                Dance Studio                    Spa   
38                        Park      Convenience Store   
42            Business Service               Bus Line   
43        Gym / Fitness Center       Department Store   
61                Home Service                 Garden   
62                        Park          Jewelry Store   
70                      Bakery      Convenience Store   
77  Construction & Landscaping                 Bakery   
88                        Pool                  River   
91                      Bakery     Miscellaneous Shop   
97                        Park               Bus Line   

        4th Most Common Venue      5th Most Common Venue  \
7              Ice Cream Shop                       Park   
11  Middle Eastern Restaurant             Sandwich Place   
14      Performing Arts Venue                      Motel   
23  Middle Eastern Restaurant        Monument / Landmark   
29              Shopping Mall                      Hotel   
34                       Park               Intersection   
38         Miscellaneous Shop                      Motel   
42                Swim School          Accessories Store   
43               Dance Studio                      Hotel   
61             Medical Center   Mediterranean Restaurant   
62                      Trail          Accessories Store   
70                       Park        Japanese Restaurant   
77                       Park         Miscellaneous Shop   
88          Accessories Store  Middle Eastern Restaurant   
91        Moroccan Restaurant        Monument / Landmark   
97          Mobile Phone Shop         Miscellaneous Shop   

              6th Most Common Venue            7th Most Common Venue  \
7                     Metro Station                     Intersection   
11                Accessories Store       Modern European Restaurant   
14                   Massage Studio                   Medical Center   
23  Molecular Gastronomy Restaurant       Modern European Restaurant   
29                Accessories Store               Miscellaneous Shop   
34                       Beer Store                      Curling Ice   
38              Moroccan Restaurant              Monument / Landmark   
42                Mobile Phone Shop              Monument / Landmark   
43                Food & Drink Shop                   Sandwich Place   
61                   Massage Studio                      Men's Sto

Cluster 5

In [37]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  Scarborough               4  Fast Food Restaurant     Accessories Store   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0  Middle Eastern Restaurant   Moroccan Restaurant   Monument / Landmark   

             6th Most Common Venue       7th Most Common Venue  \
0  Molecular Gastronomy Restaurant  Modern European Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0     Mobile Phone Shop    Miscellaneous Shop     Mexican Restaurant

Success cluster the Toronto neighborhoods based on venue category.